Default subsets from creator plugin

davidlatwe opened the pull request:
## Feature Propose

Make TDs able to predefine some commonly used subset names in the creator plugin.

Example creator:
```python
import avalon.maya

class ModelCreator(avalon.maya.Creator):
    name = "modelDefault"
    label = "model"
    family = "pipeline.model"
    subsets = [
        "HighPoly",
        "LowPoly",
    ]
```

Those predefined subset names will be listed in the Creator App's subset menu for artists to pick up, along with previous published subset names.

davidlatwe commented:
> though the `subsets` attribute is somewhat confusing.

This might relate to the concept of #269 :
*the subset name = family root name + variant name*

So how about renaming the attribute `subsets` to `variants` ?
That way, the `name` will still being as a default subset name, which precomposed without from family and default variant.

BigRoy commented:
The issue I'm having is that the name is also listed among the "variants". I'd say drop `name` and just do:

```python
subsets = ["default", "highPoly", "lowPoly"]
```

The first one in the list will be just the one that's displayed by default. I believe the name is not used anywhere else than to base the "default subset name" of. Should we maybe deprecate name (allow it to remain `None` - and maybe show a deprecation message?) and support solely passing `subsets` or `variants` or `names`?

- Tagging @mkolar as he might have some input too.

mkolar commented:
> The first one in the list will be just the one that's displayed by default.

I agree with this completely. having name in there feels like extra information for no reason. And is in danger of staying hovering around in the code similar to how `family` and `families`  are both used in pyblish still even though family is practically not needed.

BigRoy commented on avalon/pipeline.py:
```diff
@@ -216,24 +216,40 @@ def remove(self, container):
+    variants = None
 
     def __init__(self, name, asset, options=None, data=None):
-        self.name = name or self.name
```
Removing `self.name` might introduce backwards compatibility issues - since other code could be using it. E.g. we've been using it [here](https://github.com/Colorbleed/colorbleed-config/blob/master/colorbleed/plugins/maya/create/colorbleed_renderglobals.py#L54)

How about we remove it as attribute like you did (line 219) but still assign name to `self.name` in `__init__`?

davidlatwe replied:
This has been take cared in 430d364

BigRoy commented:
The visual changes I like, so the GUI seems fine. However, by avoiding `Creator.name` and just supporting the multiple of recommended subsets I was hoping that the code itself would somewhat simplify - as name felt a bit odd to implement on the Creator.

mkolar commented on avalon/pipeline.py:
```diff
@@ -216,24 +216,40 @@ def remove(self, container):
-    name = None
     label = None
     family = None
+    variants = None
```
I quite like that definition where `subste` = `family` + `variant` we use it that way and talk about it that way in the studio, but i never though of specifying it like the. @mottosso what do you think?

BigRoy replied:
Note that Variants was coined before, see: https://github.com/getavalon/core/issues/269

davidlatwe replied:
I think I would pick `defaults` for now, since it's indeed the default names for subsets, and would also make sense if one describe it as *default variants*.

mkolar replied:
> Variants was coined before

right. that's because we use it internally among artists :)

> I think I would pick defaults for now

that works nicely I don't mind either way.

davidlatwe commented:
Sorry, this actually should be in another issue or PR, I'd better revert that commit 2cf4f52 for now :/

BigRoy commented:
Thanks @davidlatwe - looks pretty good to me.

davidlatwe commented:
I have added one last commit for ensuring capitalize subset name and ignoring letter case.
Ready to merge !

davidlatwe commented:
> It should only uppercase the first character, right?

Oh yes, you are right !
I missed that use case.

I was meant to prevent the possible subset name duplication in the menu due to different letter cases. Since the subset name parsed from database would have first letter uppercased, and if the `defaults` from the creator plugin were first letter lowercased, the menu will have duplicated names.

I have reverted the last commit and changed to another implementation.

davidlatwe commented:
Merging this if no any further doubts or objections :)
Thank you all !

davidlatwe commented:
Merged :)

BigRoy commented:
Just wanted to let you know that this PR introduced an issue for us for plug-ins that create instances that weren't intended to be picked up by the default collector. Now with this PR, it feels like a hack because it's broken so it's likely a bit hacky on our end. ;)

For example these plug-ins:
- https://github.com/Colorbleed/colorbleed-config/blob/72b8aa5cd0612b8a80ad4afaa70efd53bfcd821a/colorbleed/plugins/maya/create/colorbleed_vrayscene.py
- https://github.com/Colorbleed/colorbleed-config/blob/72b8aa5cd0612b8a80ad4afaa70efd53bfcd821a/colorbleed/plugins/maya/create/colorbleed_renderglobals.py

Note how they are popping the [`subset` data](https://github.com/Colorbleed/colorbleed-config/blob/72b8aa5cd0612b8a80ad4afaa70efd53bfcd821a/colorbleed/plugins/maya/create/colorbleed_renderglobals.py#L33) as this particular instance did not need it. 

Any pointers on which way this could be fixed best? We'd love to maintain the ability to create this more "global" publish instances through the Creator.

davidlatwe commented:
Oh ! This seems the best option here :
> Better proposal, keep self.name usage

Yeah, that's better and workable for all of us !
